### PR TITLE
[F2F-291] Create L2 stack for DynamoDB state tables

### PR DESCRIPTION
**Pull request by darrendunford**
### What changed

Add table, keys, aliases and outputs for the session and person identity table

### Why did it change

Persistent state tables must be in independent stacks for stability

### Issue tracking

- [F2F-291](https://govukverify.atlassian.net/browse/F2F-291)

### Environment variables or secrets

- [X] No environment variables or secrets were added or changed

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

[F2F-291]: https://govukverify.atlassian.net/browse/F2F-291?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by richa-misra — DISMISSED** [review]

File: infra-l2-dynamo/package.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+    "name": "dca-mock-readid",
```
Comment (richa-misra) [inline infra-l2-dynamo/package.json:2]: dca ref

File: infra-l2-dynamo/package.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        "@types/js-yaml": "^4.0.5",
+        "esbuild-jest": "^0.5.0",
+        "jest": "^29.0.0",
+        "jest-junit": "^14.0.1",
+        "ts-node": "^10.4.0",
+        "typescript": "^4.7.3",
+        "yaml-cfn": "^0.3.1"
+    },
+    "scripts": {
+        "test": "./node_modules/.bin/jest",
```
Comment (richa-misra) [inline infra-l2-dynamo/package.json:23]: needed?

File: infra-l2-dynamo/template.yaml [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    Value: !GetAtt PersonIdentityTableKey.Arn
+    Export:
+      Name: !Sub "${AWS::StackName}-person-identity-table-key-arn"
+  PersonIdentityTableKeyId:
+    Description: "The Id of the person identity table encryption key"
+    Value: !Ref PersonIdentityTableKey
+    Export:
+      Name: !Sub "${AWS::StackName}-person-identity-table-key-id"
+  PersonIdentityTableKeyAlias:
+    Description: "The Alias of the person udentity table encryption key"
```
Comment (richa-misra) [inline infra-l2-dynamo/template.yaml:246]: nit: udentity

**Comment by richa-misra** [comment]
package-lock not needed to be merged in

File: infra-l2-dynamo/package.json (hunk 2 above)
Comment (andyloughran) [reply infra-l2-dynamo/package.json:23]: Yeah, this will be being used to execute the infra tests.

Comment (andyloughran) [reply infra-l2-dynamo/package.json:23]:
```suggestion
        "test": "npm run infra",
```

**Review by andyloughran — CHANGES_REQUESTED** [review]

**Comment by andyloughran** [comment]
Please can we add a new ticket for addung the .git/workflow files for PR and merge of this 'app' so that it can be tested and goto higher environments.

It needs excluding from the existing workflows on path.

File: infra-l2-dynamo/template.yaml [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        AttributeName: expiryDate
+        Enabled: true
+      PointInTimeRecoverySpecification:
+        PointInTimeRecoveryEnabled: true
+      SSESpecification:
+        KMSMasterKeyId: !Ref SessionTableKey
+        SSEEnabled: true
+        SSEType: KMS
+
+  SessionTableKey:
```
Comment (richa-misra) [inline infra-l2-dynamo/template.yaml:108]: Does [cic-front-sessions table](https://github.com/alphagov/di-ipv-cri-cic-front/blob/b5280ae5ac650b61999598827356a00a3d640a39/template.yaml#L632) need to be part of this stack or will that be a separate one?

**Review by andyloughran — DISMISSED** [review]

File: infra-l2-dynamo/template.yaml (hunk 4 above)
Comment (andyloughran) [reply infra-l2-dynamo/template.yaml:108]: That can be in the FE template, as it's not long-term stateful.

To the same end, the SessionTable here could have potentially lived in the backend stack (as it's not long lived) - but no harm in keeping it here now it exists.

Comment (andyloughran) [reply infra-l2-dynamo/template.yaml:108]: @darrendunford Coral FYI to comment, confirm and resolve.

Comment (andyloughran) [reply infra-l2-dynamo/template.yaml:108]: @CoralSmith

Comment (CoralSmith) [reply infra-l2-dynamo/template.yaml:108]: We did say we'd have long term stateful storage in layer 2, and this is described as that. In workshops we defined long term as 'X hours' which is _ambiguous_... so I can see why we've put this here. 

I'm fine with it staying here, I can't foresee it causing us issues.

**Review by andyloughran — APPROVED** [review]

File: .github/workflows/pull-request-Dynamo.yml [hunk 5]
```diff
@@ -0,0 +1,102 @@
+name: Pull Request CI - Dynamo Infra
+on:
+  # Triggers the workflow on push or pull request events but only for the "main" or "develop" branch
```
Comment (darrendunford) [inline .github/workflows/pull-request-Dynamo.yml:3]: should the comment mention the "develop" branch?

Comment (andyloughran) [reply .github/workflows/pull-request-Dynamo.yml:3]: PR should only merge into main - not sure why develop is here?

Comment (andyloughran) [inline .github/workflows/pull-request-Dynamo.yml:3]:
```suggestion
  # Triggers the workflow on push or pull request events but only for the "main" branch.
```

File: infra-l2-dynamo/template.yaml (hunk 4 above)
Comment (darrendunford) [reply infra-l2-dynamo/template.yaml:108]: agreed. we said that we would put all the BE tables here in this L2 stack for simplicity, recognising the BE for this CRI as a whole is long-term stateful.

File: src/tests/infra/template.infra-test.ts [hunk 6]
```diff
@@ -31,7 +31,7 @@ it('Has tracing enabled on at least one API', () => {
     })
 })
 
-it.skip('There are 1 lambdas defined, all with a specific permission:', () => {
+it('There are 1 lambdas defined, all with a specific permission:', () => {
```
Comment (darrendunford) [inline src/tests/infra/template.infra-test.ts:34]: why is this in this PR?

File: src/package.json [hunk 7]
```diff
@@ -37,7 +37,7 @@
         "ts-jest": "^29.0.5",
         "ts-node": "^10.4.0",
         "ts-standard": "^11.0.0",
-        "typescript": "^4.9.4",
+        "typescript": "^4.5.0",
```
Comment (darrendunford) [inline src/package.json:40]: why is this in this PR?

**Review by andyloughran — APPROVED** [review]

**Pull request merged**